tests/MiddleMailServiceTests.cs

```diff
@@ -107,6 +141,27 @@ public async Task ProcessingContinuesAfterGeneralProcessingException() {
 			}
 		}
 
+		[Fact]
+		public async Task RateLimitAllowsCorrectNumberThrough() {
+			for (int i = 0; i < 3; i++) {
+				var emailMessage = FakerFactory.EmailMessageFaker.Generate();
+				await rateLimitedCallback(emailMessage);
+				rateLimitedProcessorMock.Verify(m => m.ProcessAsync(It.IsAny<EmailMessage>()), Times.Exactly(i + 1));
+			}
+
+			// If you want to test that it takes the correct amount of time to delay,
+			// uncomment each commented line,
+			// and change the number of verified sent emails to 4.
+			// DateTime before = DateTime.Now;
+			var notSentMessage = FakerFactory.EmailMessageFaker.Generate();
+			await Task.WhenAny(rateLimitedCallback(notSentMessage), Task.Delay(TimeSpan.FromSeconds(5)));
```

Comment:
What part of how it's written confuses you? The concept of the test is as follows:
3 times:
- send a message
- verify that the message is delivered

one more time:
- initiate sending a message
- wait less time than the rate limit window
- verify that the message is not delivered

In other words, where should I add a comment?

Reply:
I believe it would be already good if you move the commented out things that test a separate thing into a different method and separate the tests into `Arrange / Act / Assert` sections. 
Anyway, I don't think you really need to test that the ratelimiter works, but just that it is called correctly and that you behave correctly depending on the lease (wait or process).

Reply:
Looking at this again now, I still think this test is helpful to check that the behavior works as intended, even if it doesn't fit into the canonical Arrange / Act / Assert sections. I've separated the test into two tests, one of which could be commented out to speed up manual retesting if necessary.